add ob.d and @live prototype implementation of Ownership/Borrowing

WalterBright opened the pull request:
This is far enough along that it can be experimented with.

WalterBright committed:
add ob.d and @live prototype implementation

WalterBright commented:
Both the semaphoreci and win32-ldc failures appear to be seg faults in ldc.

WalterBright commented:
@kinke any ideas on the ldc seg faults?

kinke commented:
Nope, that's definitely not enough information for me to speculate.

kinke commented:
=> https://github.com/ldc-developers/ldc/issues/3234

tgehr commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+
+It will not detect attempts to dereference `null` pointers or possibly
+`null` pointers. This is unworkable because there is no current method
+of annotating a type as a non-`null` pointer.
```
Contracts! :)

Geod24 reviewed (commented):
So what is the goal here ? When should this be considered ready to merge ? It is currently very limited, and as you mentioned, only a prototype will tell us if this can work.

I have some concerns about some of the design decisions, for example the non-transitivity (ability to call non-`@live` functions and completely bypass the checks). But I'll try to stay in the framework you built for the review.

First piece of code I tried:
```D
void main () @live
{
    int* ptr;
}
```

Does not compile because of a dangling pointer. Likewise, `int* ptr = new int` does not compile, which kind of defeat the point of having a GC.

Next I tried the following code:
```D
import core.stdc.stdlib;

void main () @live
{
    int* ptr = new int;
    foo(ptr);
}

void foo (int* ptr) @live
{
    int* g = ptr;
    free(g);
}
```

It crashed the compiler:
```
% ./dmd -run ob.d
dmd_runUzhSCf(50374,0x119a72dc0) malloc: *** error for object 0x10a495000: pointer being freed was not allocated
dmd_runUzhSCf(50374,0x119a72dc0) malloc: *** set a breakpoint in malloc_error_break to debug
Error: program killed by signal 6
```

Removing the `import` (which leads to an error) also triggers an ICE:
```console
% ./dmd -run ob.d
ob.d(14): Error: free is not defined, perhaps import core.stdc.stdlib; is needed?
---
ERROR: This is a compiler bug.
Please report it via https://issues.dlang.org/enter_bug.cgi
with, preferably, a reduced, reproducible example and the information below.
DustMite (https://github.com/CyberShadow/DustMite/wiki) can help with the reduction.
---
DMD v2.089.0-rc.1-164-ged4482fd4
predefs   DigitalMars Posix OSX CppRuntime_Clang darwin LittleEndian D_Version2 all D_SIMD D_InlineAsm_X86_64 X86_64 D_LP64 D_PIC assert D_ModuleInfo D_Exceptions D_TypeInfo D_HardFloat D_ObjectiveC
binary    ./dmd
version   v2.089.0-rc.1-164-ged4482fd4
config    dmd.conf
DFLAGS    -I./../../src/phobos -I./../../src/druntime/import -L-L./../lib
---
core.exception.AssertError@dmd/ob.d(879): Assertion failure
----------------
??:? _d_assertp [0x109b78e31]
dmd/access.d:394 void dmd.ob.toObNodes(ref dmd.root.array.Array!(dmd.ob.ObNode*).Array, dmd.statement.Statement).visit(dmd.statement.Statement) [0x1099fed6c]
dmd/access.d:394 void dmd.ob.toObNodes(ref dmd.root.array.Array!(dmd.ob.ObNode*).Array, dmd.statement.Statement) [0x1099fea28]
dmd/access.d:394 void dmd.ob.oblive(dmd.func.FuncDeclaration) [0x1099fd6b1]
dmd/access.d:394 _ZN16Semantic3Visitor5visitEP15FuncDeclaration [0x109a342a2]
dmd/access.d:394 _ZN15FuncDeclaration6acceptEP7Visitor [0x109989d15]
dmd/access.d:394 _Z9semantic3P7DsymbolP5Scope [0x109a2fc5d]
dmd/access.d:394 _ZN16Semantic3Visitor5visitEP6Module [0x109a30063]
dmd/access.d:394 _ZN6Module6acceptEP7Visitor [0x1098fea75]
dmd/access.d:394 _Z9semantic3P7DsymbolP5Scope [0x109a2fc5d]
dmd/access.d:394 int dmd.mars.tryMain(ulong, const(char)**, ref dmd.globals.Param) [0x1099e68b1]
dmd/access.d:394 _Dmain [0x1099e7ea1]
```

However it looks like the following code silence the "dangling pointer" error:
```D
void main () @safe @live
{
    int* ptr;
    foo(ptr);
}

void foo (ref int* ptr) @safe @live {}
```

It is quite hard to make sense of this TBH. I have not started to review the implementation, as I focused on test cases, but since it is quite buggy (ICE & very little language construct supports), it's hard to make sense of the intentions, for example what error is this dangling pointer error trying to prevent ?

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+
+The only pointers that are tracked are those declared in the `@live` function
+as `this`, function parameters or local variables. Variables from other
+functions are not tracked, even `@live` ones, as the analysis of interactions
```
What does "Variables from other functions are not tracked" mean?

WalterBright replied:
It means that called functions are not looked at beyond the function signature.

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+functions are not tracked, even `@live` ones, as the analysis of interactions
+with other functions depends
+entirely on that function signature, not its internals.
+Parameters that are const are not tracked.
```
Does this mean that const local variables aren't either? const this?

WalterBright replied:
`const` local pointers are tracked. `const this` is not tracked (it can't be free'd anyway, so why bother tracking it?)

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+
+Each pointer is in one of the following states:
+
+1. Undefined
```
How does a pointer get this state?

WalterBright replied:
```
T* p = void;
```
also:
```
T* p = ...;
free(p);
// now `p` is undefined
```

atilaneves replied:
I think these examples should go in the DIP at in this section.

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+
+3. assignment
+
+4. passed to an `out` function parameter (changes state after the function returns)
```
And mutable `ref`.

WalterBright replied:
ok

atilaneves commented on changelog/ob.md:
````diff
@@ -0,0 +1,282 @@
+
+```
+It doesn't seem to make much sense to have such pointers as
+`scope`, perhaps this can be resolved by making such an error.
````
I think that means that pointers that are assigned to something like `malloc`. It doesn't make sense for them to be scope, and I think that making this an error is a very good idea.

Geod24 replied:
That conflicts directly with the statement at line 15 which says that OB does not distinguish between GC and non-GC allocated memory.

WalterBright replied:
There'd be no point to `scope` attached to a GC allocated pointer, either.

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+
+### Pointer State Transitions
+
+A pointer changes its state when one of these operations is done to it:
```
It's not clear to me what the state transitions are from this description.

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+from an owner pointer, and the owner then enters the Lent state
+until after the last use of the borrowed pointer.
+
+A Borrowed pointer must have the `scope` attribute and must
```
If such a pointer isn't `scope`, what happens? Is that forbidden by the design?

WalterBright replied:
If it isn't `scope`, then it is a transfer of the source pointer to the target pointer. The state of the source pointer then becomes Undefined.

atilaneves replied:
I think that should be clarified in the description.

WalterBright replied:
I've rewritten parts of ob.md.

atilaneves commented on test/fail_compilation/fob1.d:
```diff
@@ -0,0 +1,60 @@
+@live void foo2()
+{
+    int* p;
+    p = null;
```
I'd also like to know. Pointing to null isn't dangling, pointing to stale memory is.

WalterBright replied:
`null` isn't recognized as special by the O/B system. It does not track `null` at all. This is because functions that may return `null` are not distinguishable from those that may not, rendering tracking `null` rather impotent.

atilaneves commented on test/compilable/ob1.d:
```diff
@@ -0,0 +1,25 @@
+
+@live int* foo1(int* p)
```
I think comments as to what this is testing would help comprehension. IIUC, this is testing that not reading or writing to `p` despite it supposedly being an owning pointer is not an error.

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+The pointer is in an invalid state. Dereferencing such a pointer is
+an error.
+
+2. Owner
```
What happens if an owner pointer is assigned to another owner pointer?

```
int* p = cast(int*) malloc(int.sizeof);
int* q = p;  // ???

```

WalterBright replied:
then `p` enters an undefined state.

atilaneves replied:
I think this needs to be clarified in the DIP.

WalterBright commented:
@Geod24 
> It crashed the compiler:

This has been fixed.

> Removing the import (which leads to an error) also triggers an ICE:

Also has been fixed.

WalterBright commented:
> So what is the goal here ? When should this be considered ready to merge ?

I set it up so it is fairly independent of the rest of the compiler, and is useful as a step on the way to a full implementation. If I wrote the entire thing at once, nobody would review the code, and furthermore if everyone hated it I'd have wasted much time.

So it's set up as an incremental thing. This is the initial bit, and more will follow in reviewable chunks.

WalterBright commented:
> Likewise, `int* ptr = new int;` does not compile, which kind of defeat the point of having a GC.

`@live` is quite unnecessary if you're going to solely use the GC. It's for manually managed memory. However, there is no way to distinguish a pointer to GC memory from a pointer to manually managed memory, hance `@live` treats them the same. It gives an error because `ptr` is left dangling at the end of the function.

WalterBright commented:
> what error is this dangling pointer error trying to prevent ?

```
void foo() { void* p = malloc(10); }
```
A dangling pointer is one that is never free'd. More pedantically, an "Owner" pointer needs to do something in the function that "consumes" it, and free'ing it is one way to consume it.

WalterBright commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+of aggregates. Adding support for such adds complexity,
+but does not change the nature of it, hence it is deferred to later.
+RAII objects can safely manage their own memory, so are not covered
+by OB. Whether a pointer is allocate memory using the GC or some other
```
fixed

WalterBright commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+
+It will not detect attempts to dereference `null` pointers or possibly
+`null` pointers. This is unworkable because there is no current method
+of annotating a type as a non-`null` pointer.
```
refs do not work for locals.

WalterBright commented on changelog/ob.md:
```diff
@@ -0,0 +1,282 @@
+that memory object, it can safely
+manage the memory (change its shape, allocate, free and resize) without
+pulling the rug out from under any other pointers (mutating or not)
+that may point to it.
```
Yes. If I have a pointer to const(X), and a mutating pointer to X, X can get free'd but the former will still be pointing to X.

Geod24 commented:
Thanks for answering all the questions. I checked out your branch and gave this another spin.
First, I must say that the fact that the following code does not compile is a blocker for adoption:
```D
void main () @safe @live
{
    void* ptr; // ob.d(3): Error: variable ob.main.ptr is left dangling at return
}
```
D has always been pragmatic, and introducing such idiosyncrasy brings us closer to C++, in a bad way.

> Also has been fixed.

It still crashes on undefined identifier, e.g:
```D
int* malloc();
void free(int*);
 @live void waterTight()
{
    auto p = malloc();
    scope(exit) free(p);
    pitcher();
}
```

```
ob.d(21): Error: undefined identifier pitcher
---
ERROR: This is a compiler bug.
Please report it via https://issues.dlang.org/enter_bug.cgi
with, preferably, a reduced, reproducible example and the information below.
DustMite (https://github.com/CyberShadow/DustMite/wiki) can help with the reduction.
---
DMD v2.089.0-rc.1-317-g015b69df1
predefs   DigitalMars Posix OSX CppRuntime_Clang darwin LittleEndian D_Version2 all D_SIMD D_InlineAsm_X86_64 X86_64 D_LP64 D_PIC assert D_ModuleInfo D_Exceptions D_TypeInfo D_HardFloat D_ObjectiveC
binary    ./generated/osx/release/64/dmd
version   v2.089.0-rc.1-317-g015b69df1
config    ./generated/osx/release/64/dmd.conf
DFLAGS    -I./generated/osx/release/64/../../../../../druntime/import -I./generated/osx/release/64/../../../../../phobos -L-L./generated/osx/release/64/../../../../../phobos/generated/osx/release/64 -fPIC
---
core.exception.AssertError@dmd/ob.d(960): Assertion failure
----------------
??:? _d_assertp [0x107fe8b0d]
dmd/access.d:394 void dmd.ob.insertFinallyBlockCalls(ref dmd.root.array.Array!(dmd.ob.ObNode*).Array) [0x107e7007a]
dmd/access.d:394 void dmd.ob.oblive(dmd.func.FuncDeclaration) [0x107e6d319]
dmd/access.d:394 _ZN16Semantic3Visitor5visitEP15FuncDeclaration [0x107ea35be]
dmd/access.d:394 _ZN15FuncDeclaration6acceptEP7Visitor [0x107df9e11]
dmd/access.d:394 _Z9semantic3P7DsymbolP5Scope [0x107e9ef79]
dmd/access.d:394 _ZN16Semantic3Visitor5visitEP6Module [0x107e9f37f]
dmd/access.d:394 _ZN6Module6acceptEP7Visitor [0x107d6ef49]
dmd/access.d:394 _Z9semantic3P7DsymbolP5Scope [0x107e9ef79]
dmd/access.d:394 int dmd.mars.tryMain(ulong, const(char)**, ref dmd.globals.Param) [0x107e56591]
dmd/access.d:394 _Dmain [0x107e57ba5]
```

> `@live` is quite unnecessary if you're going to solely use the GC.

The problem manifest as soon as you try to use `@live`. So mixing both GC and non-GC allocated memory is affected. We have a fair share of features that do not work together in D, I hope we don't add more.

> A dangling pointer is one that is never free'd. More pedantically, an "Owner" pointer needs to do something in the function that "consumes" it, and free'ing it is one way to consume it.

I was under the belief that the aim of this work had more to do with `@safe`ty than memory leak. Thanks for the clarification.

I *think* I start to see where you are going here. I can see why it makes sense, but it is going to require a lot of annotations to be usable.

WalterBright committed:
fix bad try-finally support

WalterBright commented:
> It still crashes on undefined identifier, e.g:

The undefined identifier wasn't the actual problem, but I fixed the actual problem anyway. (Keep the bug reports coming!)

> the fact that the following code does not compile is a blocker for adoption

Let's say I added null tracking to this PR. It would work for the case you presented. But 5 minutes later, I'd get another bug report:
```
int* oops() { return null; }
@live void foo() { int* p = oops(); }
```
about why the compiler complained about a dangling pointer. When and until D gets a notion of a non-null pointer as a type, it is useless to flow analyze null.

This PR is not at all about tracking null. It is about tracking whether a pointer has a defined value or not.

> So mixing both GC and non-GC allocated memory is affected.

Pointers to GC and non-GC memory are indistinguishable, so both must follow the @live rules in @live code. Part of Timon's proposal was to add more annotations to distinguish them.

WalterBright committed:
some clarification of the documentation

atilaneves commented on changelog/ob.md:
```diff
@@ -0,0 +1,296 @@
+to the code generated.
+If `@live` functions
+call non-`@live` functions, those called functions are expected to present
+an `@live` compatible interface, although it is not checked.
```
Given:

```d
void nonlive(int *p);
void fun() @live {
    auto p = cast(int*) malloc(int.sizeof);
    nonlive(p);
}
```

Would it be the case then under this proposal that `nonlive(p)` is considered to "consume" p?

Geod24 replied:
That was my understanding. Which makes it work with common approaches like "Library_init" and "Library_free". On the other hand, it does require a lot of annotations to work with *anything else*. And some use cases (like `getopt`) are probably not going to work (anything that requires multiple mutable pointers).

WalterBright replied:
Yes, exactly.

> it does require a lot of annotations to work with anything else. 

Don't see that.

> some use cases (like getopt) are probably not going to work (anything that requires multiple mutable pointers).

That's well known with Rust. They struggle with it a lot, which is why D allows you to dip in and out of @live as required.

atilaneves commented on test/fail_compilation/fob2.d:
```diff
@@ -0,0 +1,43 @@
+
+/* TEST_OUTPUT:
+---
+fail_compilation/fob2.d(203): Error: variable `fob2.zoo2.p` is not Owner, cannot borrow from it
```
Is this because of `p + 1`? If so, isn't the error message confusing?

WalterBright replied:
No, it's because `p` is sent twice to foo2(). `p + 1` is considered equivalent to `p` by the checker.

WalterBright replied:
BTW, I expect error messages can be improved over time. Such polishing, however, won't be in the first iteration.

12345swordy commented:
> This PR is not at all about tracking null. It is about tracking whether a pointer has a defined value or not.

Does this mean you are open for a dip that introduce an built in nullable type system?

WalterBright committed:
fix borrows through aggregate

WalterBright commented:
> Does this mean you are open for a dip that introduce an built in nullable type system?

You can certainly submit any DIP you like. But please, discuss it on the n.g., not here.

WalterBright committed:
scope pointers inited with untracked expressions are Owners

WalterBright committed:
ignore ErrorExp
